vllm/utils.py

```diff
@@ -101,14 +102,31 @@ def put(self, key: Hashable, value: T) -> None:
         self.cache.move_to_end(key)
         self._remove_old_if_needed()
 
+    def pin(self, key: Hashable) -> None:
+        if key not in self.cache:
+            raise ValueError(f"Cannot pin key: {key} not in cache.")
+        self.pinned_items.add(key)
+
+    def _unpin(self, key: Hashable) -> None:
+        self.pinned_items.remove(key)
+
     def _on_remove(self, key: Hashable, value: Optional[T]):
         pass
 
-    def remove_oldest(self):
+    def remove_oldest(self, remove_pinned=False):
         if not self.cache:
             return
-        key, value = self.cache.popitem(last=False)
-        self._on_remove(key, value)
+
+        if not remove_pinned:
+            if all(key in self.pinned_items for key in self.cache):
+                raise RuntimeError("All items are pinned, \
+                                   cannot remove oldest from the cache.")
+            # pop the oldest item in the cache that is not pinned
+            lru_key = next(key for key in self.cache
+                           if key not in self.pinned_items)
```

Comment:
```suggestion
            # pop the oldest item in the cache that is not pinned
            lru_key = next((key for key in self.cache
                           if key not in self.pinned_items), ALL_PINNED_SENTINEL)
            if lru_key is ALL_PINNED_SENTINEL:
                raise RuntimeError("All items are pinned, "
                                   "cannot remove oldest from the cache.")
```

and somewhere in the file
```
class _Sentinel:
    ...

ALL_PINNED_SENTINEL = _Sentinel()
```

Reply:
thanks for the suggestion, updated.